src/timeout.rs

```diff
@@ -0,0 +1,188 @@
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use pin_project_lite::pin_project;
+use rand::{Rng, RngCore};
+use tokio::time::{sleep_until, Instant, Sleep};
+
+// # Compliance: 5.2
+// To prevent split votes in the first place, election timeouts are
+// chosen randomly from a fixed interval (e.g., 150–300ms).
+const MIN_REARM_DURATION: u64 = 150;
+const MAX_REARM_DURATION: u64 = 300;
+
+/// A auto-rearming Timeout which can be used to make perpetual progress based on a timeout
+/// duration.
+///
+/// The `timeout_ready() -> TimeoutReady` returns a Future which must be polled to make progress.
+/// Upon expiration, TimeoutReady will auto-rearm the Timeout (relative to `Instant::now()`) to
+/// ensure perpetual progress.
+///
+/// Due to auto-rearming, a Timeout is meant to be reused even after it expires:
+///
+/// ``` none
+/// let mut prng = Pcg32::from_entropy();
+/// let mut timout = Timeout::new(&mut prng);
+///
+/// // await the first timeout
+/// pin!(timout.timeout_ready(&mut prng)).await;
+///
+/// // await the second timeout
+/// pin!(timout.timeout_ready(&mut prng)).await;
+///
+/// ```
+#[derive(Debug)]
+pub struct Timeout {
+    // A sleep future which completes after the specified duration.
+    //
+    // https://docs.rs/tokio/1.40.0/tokio/time/fn.sleep.html
+    // > Sleep operates at millisecond granularity and should not be used for tasks that require
+    // > high-resolution timers.
+    sleep: Pin<Box<Sleep>>,
+}
+
+impl Timeout {
+    /// Returns an armed Timeout.
+    fn new<R: RngCore>(prng: &mut R) -> Self {
+        let duration = Self::rearm_duration(prng);
+        let expire = Instant::now() + duration;
+        let sleep = Box::pin(sleep_until(expire));
+
+        Timeout { sleep }
+    }
+
+    /// Check if the timeout has expired.
+    fn poll_ready(&mut self, ctx: &mut Context) -> Poll<()> {
+        self.sleep.as_mut().poll(ctx)
+    }
+
+    /// Reset the expiration time.
+    ///
+    /// Sets the next timeout to a duration relative to `Instant::now()`.
+    fn rearm<R: RngCore>(&mut self, prng: &mut R) {
+        let duration = Self::rearm_duration(prng);
+        let expire = Instant::now() + duration;
+
+        // reset the sleep future
+        self.sleep.as_mut().reset(expire);
+    }
+
+    /// Randomly select a duration for the next timeout.
+    fn rearm_duration<R: RngCore>(prng: &mut R) -> Duration {
+        let range = prng.gen_range(MIN_REARM_DURATION..MAX_REARM_DURATION);
+        Duration::from_millis(range)
+    }
+
+    /// Returns the instant when the timout will expire.
+    fn expire_instant(&self) -> Instant {
+        self.sleep.deadline()
+    }
+
+    /// Returns a Future which can be polled to check if the timout has expired.
+    fn timeout_ready<'a, R: RngCore>(&'a mut self, prng: &'a mut R) -> TimeoutReady<'a, R> {
```

Comment:
Would be good to mark whatever constitues the public API as `pub(crate)` (or whatever).

Reply:
I was going to wait but it makes sense to do it now. :+1: 

I ended up moving `fn expire_instant` to testing since that the only place it was being used.